EIPS/eip-3157.md

```diff
@@ -0,0 +1,265 @@
+     * @param amounts The amount of tokens lent.
+     * @param data Arbitrary data structure, intended to contain user-defined parameters.
+     */
+    function flashLoan(
+        IERC3157BatchFlashBorrower receiver,
+        address[] calldata token,
+        uint256[] calldata amount,
+        bytes[] calldata data
+    ) external returns (bool);
+}
+```
+
+The `maxFlashAmount` function MUST return the maximum loan possible for each `token`. If a `token` is not currently supported `maxFlashAmount` MUST return 0, instead of reverting.
+
+The `flashFee` function MUST return the fees charged for each loan of `amount` `token`. If a token is not supported `flashFee` MUST revert.
+
+The `flashLoan` function MUST include a callback to the `onFlashLoan` function in a `IERC3157BatchFlashBorrower` contract.
+
+```
+function flashLoan(
+    IERC3157BatchFlashBorrower receiver,
+    address[] calldata tokens,
+    uint256[] calldata amounts,
+    bytes calldata data
+) external returns (bool) {
+  ...
+  require(
+      receiver.onFlashLoan(msg.sender, tokens, amounts, fees, data),
+      "IERC3157: Callback failed"
+  );
+  ...
+}
+```
+
+The `flashLoan` function MUST transfer `amount` of each `token` to `receiver` before the callback to the borrower.
+
+The `flashLoan` function MUST include `msg.sender` as the `sender` to `onFlashLoan`.
+
+The `flashLoan` function MUST NOT modify the `tokens`, `amounts` and `data` parameters received, and MUST pass them on to `onFlashLoan`.
+
+The `lender` MUST verify that the `onFlashLoan` callback returns the keccak256 hash of "ERC3157BatchFlashBorrower.onFlashLoan".
+
+The `flashLoan` function MUST include a `fees` argument to `onFlashLoan` with the fee to pay for each individual `token` and `amount` lent, ensuring that `fees[i] == flashFee(tokens[i], amounts[i])`.
+
+After the callback, for each `token` in `tokens`, the `batchFlashLoan` function MUST take the `amounts[i] + fees[i]` of `tokens[i]` from the `receiver`, or revert if this is not successful.
+
+If successful, `flashLoan` MUST return `true`.
+
+For all functions above, including both mandatory and optional sections, address(1) is used as a sentinel value for Ether. If the token parameter is address(1) then the function should be processed as defined except using Ether instead of a token.
+
+### Receiver Specification
+
+A `receiver` of flash loans MUST implement the IERC3157FlashBorrower interface:
+
+```
+pragma solidity ^0.7.0 || ^0.8.0;
+
+
+interface IERC3157FlashBorrower {
+
+    /**
+     * @dev Receive a flash loan.
+     * @param sender The initiator of the loan.
+     * @param token The loan currency.
+     * @param amount The amount of tokens lent.
+     * @param fee The additional amount of tokens to repay.
+     * @param data Arbitrary data structure, intended to contain user-defined parameters.
+     * @return The keccak256 hash of "ERC3157BatchFlashBorrower.onFlashLoan"
+     */
+    function onFlashLoan(
+        address sender,
```

Comment:
Consider renaming this parameter to `initiator` to add a bit more clarity.  This address doesn't actually "send" anything.

Reply:
Fixed, will do the same for ERC3156.